iceoryx_posh/test/moduletests/test_roudi_iceoyx_roudi_memory_manager.cpp

```diff
@@ -0,0 +1,139 @@
+class IceoryxRoudiMemoryManager_test : public Test
+{
+  public:
+    std::unique_ptr<IceOryxRouDiMemoryManager> m_roudiMemoryManagerTest;
+
+    void SetUp() override
+    {
+      auto config = iox::RouDiConfig_t().setDefaults();
+      m_roudiMemoryManagerTest = std::unique_ptr<IceOryxRouDiMemoryManager>(new IceOryxRouDiMemoryManager(config));
+    }
+
+    void TearDown() override
+    {
+    }
+};
+
+TEST_F(IceoryxRoudiMemoryManager_test, ConstructorSuccess)
+{
+    EXPECT_THAT(m_roudiMemoryManagerTest, Not(Eq(nullptr)));
+}
+
+TEST_F(IceoryxRoudiMemoryManager_test, IntrospectionMemoryManagerNulloptWhenNotPresent)
+{
+    auto result = m_roudiMemoryManagerTest->introspectionMemoryManager();
+    EXPECT_THAT(result, Eq(iox::cxx::nullopt_t()));
+}
+
+TEST_F(IceoryxRoudiMemoryManager_test, segmentManagerNulloptWhenNotPresent)
+{
+    auto resultTest = m_roudiMemoryManagerTest->segmentManager();
+    EXPECT_THAT(resultTest, Eq(iox::cxx::nullopt_t()));
+}
+
+TEST_F(IceoryxRoudiMemoryManager_test, portPoolNulloptWhenNotPresent)
+{
+    auto testResult = m_roudiMemoryManagerTest->portPool();
+    EXPECT_THAT(testResult, Eq(iox::cxx::nullopt_t()));
+}
+
+TEST_F(IceoryxRoudiMemoryManager_test, CreateAndAnnouceMemoryHasNoError)
+{
+    auto testResult = m_roudiMemoryManagerTest->createAndAnnounceMemory();
+    EXPECT_THAT(testResult.has_error(), Eq(false));
+}
+
+TEST_F(IceoryxRoudiMemoryManager_test, MgmtMemoryProviderReturnNonNullPtr)
+{
+    auto testResult = m_roudiMemoryManagerTest->mgmtMemoryProvider();
+    EXPECT_THAT(testResult, Not(Eq(nullptr)));
+}
+
+TEST_F(IceoryxRoudiMemoryManager_test, AcquiringIntrospectionMemoryManagerAfterCreateAndAnnounceMemoryIsSuccessful)
+{
+    auto tr = m_roudiMemoryManagerTest->createAndAnnounceMemory();
+
+    EXPECT_THAT(tr.has_error(), Eq(false));
+
+    auto result = m_roudiMemoryManagerTest->introspectionMemoryManager();
+    EXPECT_THAT(result, Not(Eq(iox::cxx::nullopt_t())));
+}
+
+TEST_F(IceoryxRoudiMemoryManager_test, AcquiringSegmentManagerAfterCreateAndAnnounceMemoryIsSuccessful)
+{
+    auto tr = m_roudiMemoryManagerTest->createAndAnnounceMemory();
+
+    EXPECT_THAT(tr.has_error(), Eq(false));
+
+    auto resultTest = m_roudiMemoryManagerTest->segmentManager();
+    EXPECT_THAT(resultTest, Not(Eq(iox::cxx::nullopt_t())));
+}
+
+TEST_F(IceoryxRoudiMemoryManager_test, AcquiringPortPoolAfterCreateAndAnnounceMemoryIsSuccessful)
+{
+    auto tr = m_roudiMemoryManagerTest->createAndAnnounceMemory();
+
+    EXPECT_THAT(tr.has_error(), Eq(false));
+
+    auto testResult = m_roudiMemoryManagerTest->portPool();
+    EXPECT_THAT(testResult, Not(Eq(iox::cxx::nullopt_t())));
+}
+
+TEST_F(IceoryxRoudiMemoryManager_test, DestroyMemoryIntrospectionMemoryManagerReturnNullOpt)
+{
+    auto testResult = m_roudiMemoryManagerTest->createAndAnnounceMemory();
+    
+    EXPECT_THAT(testResult.has_error(), Eq(false));
+
+    auto result = m_roudiMemoryManagerTest->destroyMemory();
```

Comment:
In a previous comment I mentioned that the following tests should be included:
 - `destroyMemory` is successful when memory was created (at the moment we have two tests in one)
 - `destroyMemory` fails when no memory was created
 - `destroyMemory` calling it multiple times failes after the second time when memory was created before
 - `createAndAnnounceMemory` calling it multiple times failes after the second them

In a safety critical system we have to tests every aspect, not only the good cases, also the bad and maybe misuse case. Here we would like to verify that our system is robust and the users are getting the information they require to handle the error.

Also try to always think of the pattern:
- Zero  (example: destroy memory which was not created)
- One  (example: destroy memory which was created)
- Many (example: calling createAndAnnounceMemory or destroyMemory multiple times)
- Corner Cases (example: you have a function which divides with a given parameter then you have to check division by zero)
- Limits (example: Are we able to create a roudi memory manager with a config where the memory size is 0)

Reply:
Fix tests with last commits. But, the multiples times call do not lead to an error actually. If I am calling several times createAndAnnouceMemory in a row, no error is set. Same for destroyMemory.

Reply:
I think you discovered a bug. @elBoberido what do you think? 
When you call `createAndAnnounceMemory` multiple times shouldn't you get a memory leak somehow.

Destroying non existing memory successfully would make sense for me since the outcome is still the same, the memory does not exist after the call. But here a warning could be maybe useful for the developer. What do you guys think: @elBoberido @JimmyBch ?

Reply:
Hi, actually only destroymemory seems to not set an error when called several times. CreateAndAnnouceMemory is fine :)

Reply:
@JimmyBch with every test you write think of that everything what the public API offers is allowed and your goal should be that our API explodes in our face ;) - this causes us some extra work but it is great for iceoryx.

Tests which are just working out of the box are  boring ;)

Reply:
@JimmyBch I just tried tried it in `iceoryx_roudi_components.cpp` by executing the following lines twice.
https://github.com/eclipse-iceoryx/iceoryx/blob/6b81ed743cd112411d9f430d2b9c82e1381e1598/iceoryx_posh/source/roudi/iceoryx_roudi_components.cpp#L30-L33

There was an error when `createAndAnnounceMemory` was called the second time. This is also necessary, since it would need to destroy the previously created shared memory which might already be in use.
What's your setup when no error is returned?

`destroyMemory` can be called multiple times. In the building blocks an error is created if e.g. the `PosixShmProvider::destroyMemory` is called multiple times, but in the higher layers the error is ignored if it's a `MemoryProviderError::MEMORY_NOT_AVAILABLE` since in this specific case it's not a problem.

Reply:
Actually it is fine. I just have to call it several times and I am getting an issue. It is only for destroy memory but as you mentioned it is normal.